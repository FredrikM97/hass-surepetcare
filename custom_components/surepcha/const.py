""" "Constants for the Sure Petcare integration."""

DOMAIN = "surepcha"

CONF_EMAIL = "email"
CONF_PASSWORD = "password"

TOKEN = "token"
CLIENT_DEVICE_ID = "client_device_id"
FACTORY = "surepetcare_factory"
TIMEOUT_API = 10
KEY_API = "surepetcare_api"
COORDINATOR_LIST = "coordinator_list"
COORDINATOR_DICT = "coordinator_dict"
COORDINATOR = "coordinator"
ENTRY_ID = "entry_id"
SCAN_INTERVAL = 300
POLLING_SPEED = "polling_speed"
LOCATION_INSIDE = "location_inside"
LOCATION_OUTSIDE = "location_outside"
OPTION_DEVICES = "devices"
DEVICE_OPTION = "device_option"
OPTIONS_FINISHED = "finished"
PRODUCT_ID = "product_id"
DEVICES = "devices"
NAME = "name"
