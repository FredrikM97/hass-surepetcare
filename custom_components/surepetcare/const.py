"""TODO."""

DOMAIN = "surepetcare"

CONF_EMAIL = "email"
CONF_PASSWORD = "password"

TOKEN = "token"
CLIENT_DEVICE_ID = "client_device_id"
FACTORY = "surepetcare_factory"
TIMEOUT_API = 10
KEY_API = "surepetcare_api"
COORDINATOR_LIST = "coordinator_list"
COORDINATOR_DICT = "coordinator_dict"
COORDINATOR = "coordinator"
