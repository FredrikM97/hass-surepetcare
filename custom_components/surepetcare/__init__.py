"""TODO."""

import logging
from typing import Any, List

from surepetcare.client import SurePetcareClient
from surepetcare.household import Household
from surepetcare.enums import ProductId

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import EVENT_HOMEASSISTANT_STOP, Platform
from homeassistant.core import Event, HomeAssistant, callback
from homeassistant.helpers import device_registry as dr

from .const import (
    CLIENT_DEVICE_ID,
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    FACTORY,
    KEY_API,
    TOKEN,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator

logger = logging.getLogger(__name__)

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SENSOR]


async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> bool:
    """TODO."""
    logger.info("async_setup_entry called for entry_id=%s", entry.entry_id)

    surepetcare_data: dict[str, Any] = {}
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = surepetcare_data

    client: SurePetcareClient = SurePetcareClient()
    # Try login with token+device_id, else fallback to email+password
    token = entry.data.get(TOKEN)
    device_id = entry.data.get(CLIENT_DEVICE_ID)
    email = entry.data.get("email")
    password = entry.data.get("password")
    if token and device_id:
        await client.login(token=token, device_id=device_id)
    elif email and password:
        await client.login(email=email, password=password)
    else:
        raise Exception("No valid credentials found in config entry data")
    surepetcare_data[FACTORY] = client

    async def on_hass_stop(event: Event) -> None:
        """Close connection when hass stops."""
        await client.close()

    # Setup listeners
    entry.async_on_unload(
        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
    )
    try:
        households: List[Household] = await client.api(Household.get_households())
        entities = []
        for household in households:
            entities.extend(await client.api(household.get_pets()))
            entities.extend(await client.api(household.get_devices()))
            # Add each hub device to hubs list
        await client.close()
    except Exception as exc:
        await client.close()
        raise Exception("Configuration not finished") from exc

    remove_stale_devices(hass, entry, entities)
    # Setup the device coordinators
    coordinator_data = {
        KEY_API: client,
        COORDINATOR_DICT: {},
    }

    for device in entities:
        if device.product != ProductId.HUB:
            continue
        entities.remove(device)
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    for device in entities:
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    surepetcare_data[COORDINATOR] = coordinator_data

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        surepetcare_data = hass.data[DOMAIN].pop(entry.entry_id)
        client: SurePetcareClient = surepetcare_data[FACTORY]
        await client.close()

    return unload_ok


@callback
def remove_stale_devices(
    hass: HomeAssistant, config_entry: ConfigEntry, devices: list[Any]
) -> None:
    """Remove stale devices from device registry. TODO: Work in progress and not functional yet"""

    device_registry = dr.async_get(hass)
    device_entries = dr.async_entries_for_config_entry(
        device_registry, config_entry.entry_id
    )
    all_device_ids = {str(device.id) for device in devices}
    for device_entry in device_entries:
        device_id: str | None = None  # Only define here
        # Check that device part of DOMAIN
        for identifier in device_entry.identifiers:
            if identifier[0] != DOMAIN:
                continue

            _id = identifier[1]
            device_id = str(_id)
        if device_id is None or device_id not in all_device_ids:
            logger.info(
                "Removing stale device entry %s for config entry %s",
                device_entry.id,
                config_entry.entry_id,
            )
            device_registry.async_update_device(
                device_entry.id, remove_config_entry_id=config_entry.entry_id
            )
