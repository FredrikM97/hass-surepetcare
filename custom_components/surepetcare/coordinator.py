from datetime import timedelta
import logging
from typing import Any

from surepetcare.client import SurePetcareClient
from surepetcare.devices.device import SurepyDevice

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

logger = logging.getLogger(__name__)

SCAN_INTERVAL = 300


class SurePetCareDeviceDataUpdateCoordinator(DataUpdateCoordinator[Any]):
    """Coordinator to manage data for a specific SurePetCare device."""

    config_entry: ConfigEntry

    def __init__(
        self,
        hass: HomeAssistant,
        config_entry: ConfigEntry,
        client: SurePetcareClient,
        device: SurepyDevice,
    ) -> None:
        """Initialize device coordinator."""
        self.client = client
        self.device = device
        self._exception: Exception | None = None

        super().__init__(
            hass,
            logger,
            config_entry=config_entry,
            name=f"Update coordinator for {device}",
            update_interval=timedelta(seconds=SCAN_INTERVAL),
        )

    @callback
    def _observe_update(self, device: Any) -> None:
        """Update the coordinator for a device when a change is detected."""
        self.async_set_updated_data(data=device)

    async def _async_update_data(self) -> Any:
        """Fetch data from the api for a specific device."""
        return await self.client.api(self.device.refresh())
