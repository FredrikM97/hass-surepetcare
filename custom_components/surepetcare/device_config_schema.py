"""TODO."""

from voluptuous import Required

from surepetcare.enums import ProductId

DEVICE_CONFIG_SCHEMAS = {
    ProductId.FEEDER_CONNECT: {
        "schema": {
            Required("location_inside"): str,
            Required("location_outside"): str,
        },
    },
    ProductId.DUAL_SCAN_PET_DOOR: {
        "schema": {
            Required("location_inside"): str,
            Required("location_outside"): str,
        },
    },
    ProductId.HUB: {
        "schema": None,
    },
    ProductId.PET: {
        "schema": None,
    },
}
