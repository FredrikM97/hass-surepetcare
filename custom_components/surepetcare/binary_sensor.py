"""Support for Sure Petcare binary sensors."""

from dataclasses import dataclass
from datetime import datetime
import logging
from typing import cast

from surepcio.enums import ProductId
from surepcio.devices.device import SurePetCareBase
from surepcio.devices.dual_scan_connect import Curfew
from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity import EntityCategory
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import COORDINATOR, COORDINATOR_DICT, DOMAIN, KEY_API
from .coordinator import SurePetCareDeviceDataUpdateCoordinator
from .entity import (
    SurePetCareBaseEntity,
    SurePetCareBaseEntityDescription,
)

logger = logging.getLogger(__name__)


def _next_enabled_future_curfew(device: SurePetCareBase, r: ConfigEntry) -> bool | None:
    curfews: list[Curfew] | Curfew = device.control.curfew

    if curfews is None:
        return None

    if not isinstance(curfews, list):
        curfews = [curfews]

    now = datetime.now().time()
    return any(c.enabled and c.lock_time <= now <= c.unlock_time for c in curfews)


@dataclass(frozen=True, kw_only=True)
class SurePetCareBinarySensorEntityDescription(
    SurePetCareBaseEntityDescription, BinarySensorEntityDescription
):
    """Describes SurePetCare.binary_sensor entity."""


SENSOR_DESCRIPTIONS_AVAILABLE: tuple[SurePetCareBinarySensorEntityDescription, ...] = (
    SurePetCareBinarySensorEntityDescription(
        key="connectivity",
        translation_key="connectivity",
        device_class=BinarySensorDeviceClass.CONNECTIVITY,
        field="available",
    ),
)

SENSORS: dict[str, tuple[SurePetCareBinarySensorEntityDescription, ...]] = {
    ProductId.FEEDER_CONNECT: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        *SENSOR_DESCRIPTIONS_AVAILABLE,
    ),
    ProductId.DUAL_SCAN_CONNECT: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        SurePetCareBinarySensorEntityDescription(
            key="curfew",
            translation_key="curfew_active",
            field_fn=_next_enabled_future_curfew,
            extra_field="control.curfew",
        ),
        *SENSOR_DESCRIPTIONS_AVAILABLE,
    ),
    ProductId.PET_DOOR: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        SurePetCareBinarySensorEntityDescription(
            key="curfew",
            translation_key="curfew_active",
            field_fn=_next_enabled_future_curfew,
            extra_field="control.curfew",
        ),
    ),
    ProductId.DUAL_SCAN_PET_DOOR: (*SENSOR_DESCRIPTIONS_AVAILABLE,),
    ProductId.HUB: (*SENSOR_DESCRIPTIONS_AVAILABLE,),
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up a Surepetcare config entry."""
    coordinator_data = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]
    client = coordinator_data[KEY_API]

    entities = []
    for device_coordinator in coordinator_data[COORDINATOR_DICT].values():
        descriptions = SENSORS.get(device_coordinator.product_id, ())
        entities.extend(
            [
                SurePetCareBinarySensor(
                    device_coordinator,
                    client,
                    description=description,
                )
                for description in descriptions
            ]
        )
    async_add_entities(entities, update_before_add=True)


class SurePetCareBinarySensor(SurePetCareBaseEntity, BinarySensorEntity):
    """The platform class required by Home Assistant."""

    entity_description: SurePetCareBinarySensorEntityDescription

    def __init__(
        self,
        device_coordinator: SurePetCareDeviceDataUpdateCoordinator,
        client,
        description: SurePetCareBinarySensorEntityDescription,
    ) -> None:
        """Initialize a SurePetCare binary sensor."""
        super().__init__(
            device_coordinator=device_coordinator,
            client=client,
        )

        self.entity_description = description

        self._attr_unique_id = f"{self._attr_unique_id}-{description.key}"

    @property
    def is_on(self) -> bool | None:
        """Return true if the binary sensor is on."""
        return cast(bool, self._convert_value())
