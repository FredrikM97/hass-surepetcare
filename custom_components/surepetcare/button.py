"""Support for Sure Petcare Button."""

from dataclasses import dataclass
import logging
from typing import Any
from surepcio.enums import ProductId, HubPairMode
from surepcio import SurePetcareClient
from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
from homeassistant.helpers.entity import EntityCategory
from custom_components.surepetcare.helper import MethodField, should_add_entity


from .const import (
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    KEY_API,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator
from .entity import (
    SurePetCareBaseEntity,
    SurePetCareBaseEntityDescription,
)

logger = logging.getLogger(__name__)

@dataclass(frozen=True, slots=True)
class ButtonMethodField(MethodField):
    """MethodField for button-like entities, supporting on mapping."""

    on: Any = True
    def set(self, device: object, config: dict, value: Any) -> Any:
        if value is True and self.on:
            value = self.on
        return MethodField.set(self, device, config, value)



@dataclass(frozen=True, kw_only=True)
class SurePetCareButtonEntityDescription(
    SurePetCareBaseEntityDescription, ButtonEntityDescription
):
    """Describes SurePetCare button entity."""


BUTTONS: dict[str, tuple[SurePetCareButtonEntityDescription, ...]] = {    
    ProductId.HUB: (
         SurePetCareButtonEntityDescription(
            key="pairing_mode",
            translation_key="pairing_mode",
            field=ButtonMethodField(
                path="control.pairing_mode", on=HubPairMode.ON
            ),
            entity_category=EntityCategory.CONFIG,
        ),
    )
    
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up SurePetCare sensors for each matching device."""
    coordinator_data = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]
    client = coordinator_data[KEY_API]

    entities = []
    for device_coordinator in coordinator_data[COORDINATOR_DICT].values():
        descriptions = BUTTONS.get(device_coordinator.product_id, ())
        entities.extend(
            [
                SurePetCareButton(
                    device_coordinator,
                    client,
                    description=description,
                )
                for description in descriptions
                if should_add_entity(
                    description, device_coordinator.data, config_entry.options
                )
            ]
        )
    async_add_entities(entities, update_before_add=True)


class SurePetCareButton(SurePetCareBaseEntity, ButtonEntity):
    """The platform class required by Home Assistant."""

    entity_description: SurePetCareButtonEntityDescription

    def __init__(
        self,
        device_coordinator: SurePetCareDeviceDataUpdateCoordinator,
        client: SurePetcareClient,
        description: SurePetCareButtonEntityDescription,
    ) -> None:
        """Initialize a Surepetcare sensor."""
        super().__init__(
            device_coordinator=device_coordinator,
            client=client,
        )
        self.entity_description = description
        self._attr_unique_id = f"{self._attr_unique_id}-{description.key}"

    async def async_press(self) -> None:
        """Press the button."""
        await self.send_command(1)
