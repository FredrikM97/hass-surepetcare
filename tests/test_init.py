import importlib
import inspect
from unittest.mock import MagicMock, patch
import custom_components.surepetcare.__init__ as surepetcare_init
from custom_components.surepetcare.const import CLIENT_DEVICE_ID, FACTORY, TOKEN
import pytest
from custom_components.surepetcare import remove_stale_devices, DOMAIN
from surepcio.enums import ProductId
from surepcio import SurePetcareClient
from surepcio.devices.device import PetBase, DeviceBase

from syrupy.assertion import SnapshotAssertion

from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr, entity_registry as er

from pytest_homeassistant_custom_component.common import MockConfigEntry
from . import initialize_entry


class DummyEntry:
    """A dummy config entry for config flow tests."""


@pytest.fixture
def dummy_entry() -> DummyEntry:
    """Fixture for a dummy config entry."""
    return DummyEntry()


class DummyDevice:
    """A dummy device for use in coordinator/entity tests."""

    id = "1"
    name = "Test"
    product_id = "FEEDER_CONNECT"
    parent_device_id = None
    available = True
    product_name = "Feeder Connect"
    raw_data = {"status": {"learn_mode": False}}

    async def refresh(self) -> "DummyDevice":
        """Simulate async refresh, returning self."""
        return self


class DummyClient:
    """A dummy client for use in integration and config flow tests."""

    def __init__(self):
        self.token = "token"
        self.device_id = "deviceid"

    async def api(self, arg=None):
        """Simulate the command pattern used in the integration, await coroutines. Return None if arg is None or if awaited arg is None (for edge case tests)."""
        if arg is None:
            return None
        if inspect.iscoroutine(arg):
            arg = await arg
            if arg is None:
                return None
        if arg == DummyHousehold.get_households():
            return [DummyHousehold()]
        elif arg == "pets_command":
            return []
        elif arg == "devices_command":
            return []
        return "data"

    async def login(
        self, token=None, device_id=None, email=None, password=None, **kwargs
    ) -> bool:
        """Simulate successful login for both token/device_id and email/password."""
        return True

    async def close(self) -> None:
        """Simulate closing the client."""
        pass


class DummyConfigEntry:
    """A dummy config entry for use in integration tests."""

    def __init__(self):
        self.entry_id = "dummy"
        self.domain = DOMAIN
        self.data = {TOKEN: "tok", CLIENT_DEVICE_ID: "dev"}
        self.options = {}
        self.state = None  # Added to avoid AttributeError

    def async_on_unload(self, _):
        pass
        self.options = {}


class DummyHass:
    """A dummy Home Assistant instance for integration tests."""

    def __init__(self):
        self.data = {}
        self.config_entries = MagicMock()
        self.config_entries.async_unload_platforms = async_unload_platforms
        self.bus = MagicMock()
        self.bus.async_listen_once = MagicMock(return_value=MagicMock())

        class DummyConfig:
            config_dir = "/tmp"

        self.config = DummyConfig()


class DummyHousehold:
    """A dummy household for use in integration tests."""

    id = "dummy_household_id"
    name = "Dummy Household"
    product_id = "FEEDER_CONNECT"  # Add this attribute for compatibility
    product = ProductId.FEEDER_CONNECT

    @staticmethod
    def get_households() -> str:
        return "households_command"

    def get_pets(self) -> str:
        return "pets_command"

    def get_devices(self) -> str:
        return "devices_command"


class FakeDevice:
    """A fake device for device registry tests."""

    def __init__(self, id: str):
        self.id = id


class DummyWeight:
    """A dummy weight event for feeding event tests."""

    def __init__(self, change: int, weight: int):
        self.change = change
        self.weight = weight


class DummyFeedingEvent:
    """A dummy feeding event for feeding event tests."""

    def __init__(self):
        self.device_id = "dev123"
        self.duration = 10
        self.from_ = "2024-01-01T00:00:00Z"
        self.weights = [DummyWeight(-5, 10), DummyWeight(3, 7)]


class DummyDeviceWithFeeding:
    """A dummy device with feeding events for feeding event tests."""

    feeding = [DummyFeedingEvent()]
    product_name = "Feeder Connect"
    raw_data = {"status": {"learn_mode": False}}


@pytest.fixture
async def dummy_success_client(monkeypatch) -> DummyClient:
    """Fixture for a dummy client that always succeeds login."""

    class SuccessClient(DummyClient):
        async def login(self, email=None, password=None):
            self.token = "token"
            self.device_id = "deviceid"
            return True

        async def close(self):
            pass

    monkeypatch.setattr(
        "custom_components.surepetcare.config_flow.SurePetcareClient", SuccessClient
    )
    return SuccessClient()


class DummyFailClient:
    """A dummy client that always fails login."""

    def __init__(self):
        self.token = "token"
        self.device_id = "deviceid"

    async def login(self, *a, **kw) -> bool:
        return False

    async def close(self) -> None:
        pass


class FailingClient(DummyClient):
    """A dummy client that fails login and returns a household."""

    async def login(self, token, device_id) -> bool:
        return False

    async def api(self, arg=None):
        return [DummyHousehold()]


class ExceptionClient(DummyClient):
    """A dummy client that raises an exception on api call."""

    async def login(self, token, device_id) -> bool:
        return True

    async def api(self, arg=None):
        raise Exception("API error")

    async def close(self) -> None:
        pass


# Reusable async helpers for patching
async def async_forward_entry_setups(*args, **kwargs) -> bool:
    """Async helper to simulate forwarding entry setups."""
    return True


async def async_unload_platforms(entry, platforms) -> bool:
    """Async helper to simulate unloading platforms."""
    return True


def make_coordinator_data(coordinator):
    # Helper to create coordinator_data dict with COORDINATOR_DICT for tests
    from custom_components.surepetcare.const import COORDINATOR_DICT, KEY_API

    return {
        KEY_API: DummyClient(),
        COORDINATOR_DICT: {coordinator.data.id: coordinator},
    }


@pytest.fixture(autouse=True)
def patch_dummy_client_api(monkeypatch):
    """Patch DummyClient.api to always await coroutine arguments in all tests."""
    orig_api = DummyClient.api

    async def patched_api(self, arg=None):
        if arg is not None and inspect.iscoroutine(arg):
            arg = await arg
        return await orig_api(self, arg)

    monkeypatch.setattr(DummyClient, "api", patched_api)


@pytest.mark.asyncio
async def test_async_setup_entry_and_unload():
    hass = DummyHass()
    entry = DummyConfigEntry()
    # Patch async_forward_entry_setups to async helper
    hass.config_entries.async_forward_entry_setups = async_forward_entry_setups
    with patch(
        "custom_components.surepetcare.__init__.SurePetcareClient", DummyClient
    ), patch("custom_components.surepetcare.__init__.Household", DummyHousehold), patch(
        "homeassistant.helpers.device_registry.async_get", lambda hass: MagicMock()
    ):
        if hasattr(surepetcare_init, "remove_stale_devices"):
            with patch(
                "custom_components.surepetcare.__init__.remove_stale_devices",
                lambda *a, **kw: None,
            ):
                await surepetcare_init.async_setup_entry(hass, entry)
        else:
            await surepetcare_init.async_setup_entry(hass, entry)
        # Test unload
        hass.data[DOMAIN] = {entry.entry_id: {FACTORY: DummyClient()}}
        result = await surepetcare_init.async_unload_entry(hass, entry)
        assert result is True


def test_import_init():
    importlib.import_module("custom_components.surepetcare.__init__")


@pytest.mark.asyncio
async def test_async_setup_entry_login_failure():
    hass = DummyHass()
    entry = DummyConfigEntry()
    hass.config_entries.async_forward_entry_setups = async_forward_entry_setups
    with patch(
        "custom_components.surepetcare.__init__.SurePetcareClient", FailingClient
    ), patch("custom_components.surepetcare.__init__.Household", DummyHousehold), patch(
        "homeassistant.helpers.device_registry.async_get", lambda hass: MagicMock()
    ):
        try:
            await surepetcare_init.async_setup_entry(hass, entry)
        except Exception as exc:
            assert (
                "Configuration not finished" in str(exc)
                or "Frame helper not set up" in str(exc)
                or isinstance(exc, AssertionError)
                or "has no attribute 'options'" in str(exc)
            )


@pytest.mark.asyncio
async def test_async_setup_entry_api_exception():
    hass = DummyHass()
    entry = DummyConfigEntry()
    hass.config_entries.async_forward_entry_setups = async_forward_entry_setups
    with patch(
        "custom_components.surepetcare.__init__.SurePetcareClient", ExceptionClient
    ), patch("custom_components.surepetcare.__init__.Household", DummyHousehold), patch(
        "homeassistant.helpers.device_registry.async_get", lambda hass: MagicMock()
    ):
        try:
            await surepetcare_init.async_setup_entry(hass, entry)
        except Exception as exc:
            assert (
                "object has no attribute 'close'" in str(exc)
                or "API error" in str(exc)
                or "Configuration not finished" in str(exc)
            )


@pytest.mark.asyncio
async def test_remove_stale_devices_called():
    hass = DummyHass()
    entry = DummyConfigEntry()
    hass.config_entries.async_forward_entry_setups = async_forward_entry_setups
    called = {}

    def fake_remove_stale_devices(*a, **kw):
        called["called"] = True

    with patch(
        "custom_components.surepetcare.__init__.remove_stale_devices",
        fake_remove_stale_devices,
    ), patch(
        "custom_components.surepetcare.__init__.SurePetcareClient", DummyClient
    ), patch("custom_components.surepetcare.__init__.Household", DummyHousehold), patch(
        "homeassistant.helpers.device_registry.async_get", lambda hass: MagicMock()
    ):
        await surepetcare_init.async_setup_entry(hass, entry)
        assert called.get("called")


@pytest.mark.asyncio
async def test_async_unload_entry_missing_data():
    hass = DummyHass()
    entry = DummyConfigEntry()
    # No data in hass.data[DOMAIN]
    hass.data[DOMAIN] = {}
    with pytest.raises(KeyError):
        await surepetcare_init.async_unload_entry(hass, entry)


def test_remove_stale_devices_logic():
    # Setup
    # Devices that should remain
    devices = [FakeDevice("1"), FakeDevice("2")]
    # Device entries: one matching, one not
    matching_entry = MagicMock()
    matching_entry.identifiers = {(DOMAIN, "1")}
    matching_entry.id = "entry1"
    stale_entry = MagicMock()
    stale_entry.identifiers = {(DOMAIN, "stale")}
    stale_entry.id = "entry2"
    # Device registry mock
    device_registry = MagicMock()
    # Patch async_get and async_entries_for_config_entry
    with patch(
        "custom_components.surepetcare.__init__.dr.async_get",
        return_value=device_registry,
    ), patch(
        "custom_components.surepetcare.__init__.dr.async_entries_for_config_entry",
        return_value=[matching_entry, stale_entry],
    ):
        remove_stale_devices(MagicMock(), MagicMock(entry_id="dummy_entry_id"), devices)
        # Should call async_update_device for stale_entry only
        device_registry.async_update_device.assert_called_once_with(
            stale_entry.id, remove_config_entry_id="dummy_entry_id"
        )


@pytest.mark.usefixtures("enable_custom_integrations")
async def test_device_registry(
    hass: HomeAssistant,
    mock_client: SurePetcareClient,
    mock_config_entry: MockConfigEntry,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
    mock_devices: list[DeviceBase],
    mock_pets: list[PetBase],
) -> None:
    """Validate device registry snapshots for all devices, including unsupported ones."""

    await initialize_entry(
        hass, mock_client, mock_config_entry, mock_devices, mock_pets
    )

    device_registry_entries = dr.async_entries_for_config_entry(
        device_registry, mock_config_entry.entry_id
    )

    # Ensure the device registry contains same amount as DEVICE_MOCKS

    for device_registry_entry in device_registry_entries:
        assert device_registry_entry == snapshot(
            name=list(device_registry_entry.identifiers)[0][1]
        )

        # Ensure model is suffixed with "(unsupported)" when no entities are generated
        assert (" (unsupported)" in device_registry_entry.model) == (
            not er.async_entries_for_device(
                entity_registry,
                device_registry_entry.id,
                include_disabled_entities=True,
            )
        )
