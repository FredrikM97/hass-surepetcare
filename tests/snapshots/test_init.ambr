# serializer version: 1
# name: test_device_registry[1299453]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'surepetcare',
        '1299453',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'SurePetCare',
    'model': 'DUAL_SCAN_CONNECT',
    'model_id': 6,
    'name': 'Katzenklappe',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_device_registry[222222]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'surepetcare',
        '222222',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'SurePetCare',
    'model': 'PET',
    'model_id': 0,
    'name': 'Pet with empty status',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_device_registry[269654]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'surepetcare',
        '269654',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'SurePetCare',
    'model': 'FEEDER_CONNECT',
    'model_id': 4,
    'name': 'Maui matl√•da (old) ',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': <ANY>,
  })
# ---
# name: test_device_registry[295972]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'surepetcare',
        '295972',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'SurePetCare',
    'model': 'HUB',
    'model_id': 1,
    'name': 'Hem-hub',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_device_registry[472721]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'surepetcare',
        '472721',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'SurePetCare',
    'model': 'PET',
    'model_id': 0,
    'name': 'Maui',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_device_registry[727608]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'surepetcare',
        '727608',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'SurePetCare',
    'model': 'PET_DOOR',
    'model_id': 3,
    'name': 'Laundry Door',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_device_registry[734730]
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'config_entries_subentries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'surepetcare',
        '734730',
      ),
    }),
    'labels': set({
    }),
    'manufacturer': 'SurePetCare',
    'model': 'POSEIDON_CONNECT',
    'model_id': 8,
    'name': 'Poseidon connect',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
